### ISSUE TITLE
Check current editor buffer

### ISSUE THREAD

**Issue by tonyxiao** [facebook/nuclide#604]
Atom's linter plugin (with atom-eslint) would automatically check the currently editing file buffer and show errors even before file is saved to disk. However Nuclide's issues hints only only show up after file is saved to disk. (I originally thought it was an issue related to flow, but using linter-eslint with Nuclide also exhibit the same issue).

With Nuclide (Notice the blue dot indicating file is not yet saved)
![image](https://cloud.githubusercontent.com/assets/696842/16707612/3c456c5a-45d5-11e6-8901-b2c8cd3c16b1.png)

With atom linter (Notice the blue dot indicating file is not yet saved)
![image](https://cloud.githubusercontent.com/assets/696842/16707614/4a1f60c4-45d5-11e6-9957-ecb3e1d13699.png)

**Comment by tonyxiao** [comment]
FWIW i just found this package. https://github.com/nmn/linter-flow-plus

**Comment by tonyxiao** [comment]
Just to clarify, the reason this is useful is that when developing with live reload / hot reload enabled, saving file would trigger the app to reload and state to be lost. So saving a file just to check flow / lint errors (other than the extra keystroke) is not ideal because it causes interruption in development context and might require one to navigate in the UI back to the original state again.

**Comment by zertosh** [comment]
There's an option in Nuclide's settings to enable "lint-on-the-fly" for legacy linters. But this is something don't offer for Flow (I suspect) because constantly polling flow isn't a good experience perf-wise. @nmote is this correct?

**Comment by tonyxiao** [comment]
Why is `lint-on-the-fly` only for legacy linters? What's the diff between legacy linter and linter anyways?

Recently I started using https://atom.io/packages/linter-flow with atom's linter package to get on-the-fly linting support and it seems to work pretty well.

**Comment by nmote** [comment]
Oh, I don't know why the default for that setting is false. It should be true. There isn't a performance concern. Plugins that don't perform well can opt out of linting on the fly. Anyway, in your case, checking the box should fix it until we push out the change in defaults. I just tried it out with `linter-eslint` and it worked fine.

We have our own interface for `linter`-like things, called diagnostics. The term "legacy" was just used to differentiate between the two (I think it was a poor choice of word and I will change it).

Our diagnostics interface doesn't poll providers, like the linter interface does. Instead, providers can push messages at any time. This means that it is up to each provider to choose whether to "lint on the fly" or not -- it can't be controlled by one central setting.

We used to have an option in the `nuclide-flow` package to allow this. I removed it because people kept getting confused and reporting bugs with it. Here's why:

There are two Flow commands for getting errors. `flow status` looks at the state on disk and returns all of the errors for the project. We want to be able to display all errors, not just the ones in your current file, so we need to call this at some point. `flow check-contents` takes the unsaved contents of a single file and returns its errors. Obviously, we also need to call this if we want checking without saving.

Here are the problems that I remember:
* `check-contents` will return errors even for files that don't have `/* @flow */` at the top. Since we are also calling `status` that means there will sometimes be errors and there sometimes won't be.
* `check-contents` displays errors only for the current file, as if you had the changes in the current file plus everything else on disk. That means errors will be inconsistent with other unsaved files. Granted, I'm not sure how big of an issue this is but some people I've discussed with were concerned by it.

I will talk to the Flow team and see if we can add an option to `check-contents` to respect the presence or absence of the `@flow` pragma, and then I would be okay with bringing back this feature behind a setting.

**Comment by tonyxiao** [comment]
Fantastic, thanks for the explanation @nmote . So to confirm, can I get check without save working today with flow + nuclide diagnostics?

**Comment by nmote** [comment]
No, not today. I'm getting the ball rolling on that, though. I don't want to release it until we can respect the lack of a `@flow` pragma, otherwise it can get really confusing. We got a lot of complaints about this when we allowed it.

**Comment by tonyxiao** [comment]
got it. On an unrelated note, @nmote is there a way to validate all files have `// @flow` enabled if you want to be extra strict with types? Perhaps via eslint?

**Comment by nmote** [comment]
I've added the necessary flag to Flow, and it will be available with the next release. Here's what I'm thinking for the desired behavior in Nuclide:

* Whenever you save any file, we call `flow status` (which returns all project-wide errors) and use the given results for every saved file. We will leave unsaved files alone.
* As you type, we call `flow check-contents`. We update that file only.

We still have the problem where changes in a given unsaved file are not reflected elsewhere, but I'm not sure how much of an issue that will be in practice.

**Comment by nmote** [comment]
Just put up https://github.com/facebook/flow/pull/2118 to address another issue with Flow that I found while working on this.

**Comment by STRML** [comment]
For comparison, the FlowIDE plugin for Sublime Text does a really great job of doing this quickly. It [checks for the flow pragma itself](https://github.com/tptee/FlowIDE/blob/master/flow-ide.py#L262).

I've seen it report type errors I can't get `flow` to report - seems to be an issue with context in `check-contents` - but it works really well in general.

![flow_st3](https://cloud.githubusercontent.com/assets/1197375/18134686/d5578bcc-6f64-11e6-8b8c-39a07b7b05d5.gif)

**Comment by zuhair-naqvi** [comment]
Any updates on this?

**Comment by eliseumds** [comment]
That would be a great update
